examples/recording_service/leveldb-integration/cpp/LevelDbReader.hpp

```diff
@@ -0,0 +1,262 @@
+
+    /*
+     * The return loan operation should free any resources allocated by the
+     * read() operation.
+     */
+    virtual void return_loan(
+            std::vector<dds::core::xtypes::DynamicData *>& sample_seq,
+            std::vector<dds::sub::SampleInfo *>& info_seq);
+
+    /*
+     * This method should flag Replay/Converter that all data related to the
+     * data stream has been read and that we're ready for termination.
+     */
+    virtual bool finished();
+
+    virtual void reset();
+
+private:
+
+    std::string db_filename_;
+
+    rti::routing::StreamInfo stream_info_;
+
+    dds::core::xtypes::DynamicType type_;
+
+    std::unique_ptr<leveldb::DB> data_db_;
+
+    UserDataKeyComparator key_comparator_;
+
+    std::unique_ptr<leveldb::Iterator> db_iterator_;
+
+    std::vector<char> key_buffer_;
+
+    std::vector<char> value_buffer_;
+
+    std::vector<std::shared_ptr<dds::core::xtypes::DynamicData>> loaned_samples_;
+
+    std::vector<std::shared_ptr<dds::sub::SampleInfo>> loaned_infos_;
+
+    int64_t start_timestamp_;
+
+    int64_t end_timestamp_;
+
+    int64_t current_timestamp_;
+
+    bool finished_;
+
+};
+
+/*
+ * This stream info reader implementation is able to read a Level DB discovery
+ * database recorded with the LevelDbWriter implementation, called
+ * 'DCPSPublication.dat'.
+ */
+class LevelDbStreamInfoReader :
+        public rti::recording::storage::StorageStreamInfoReader {
+public:
+
+    LevelDbStreamInfoReader(
+            const std::string& working_dir,
+            int64_t start_timestamp,
+            int64_t end_timestamp);
+
+    ~LevelDbStreamInfoReader();
+
+    /*
+     * Implementation of the read operation. It should interpret the selector
+     * state object that expresses the specific needs of Replay/Converter about
+     * the data to be provided (data not read before vs data of any kind, lower
+     * and upper time limits, etc).
+     */
+    virtual void read(
+            std::vector<rti::routing::StreamInfo *>& sample_seq,
+            const rti::recording::storage::SelectorState& selector);
+
+    /*
+     * The return loan operation should free any resources allocated by the
+     * read() operation.
+     */
+    virtual void return_loan(
+            std::vector<rti::routing::StreamInfo *>& sample_seq);
+
+    /*
+     * An int64-represented time-stamp (in nanoseconds) representing the
+     * starting point in time where recorded data exists, or when the service
+     * started executing.
+     */
+    virtual int64_t service_start_time();
+
+    /*
+     * An int64-represented time-stamp (in nanoseconds) representing the
+     * final point in time where recorded data exists, or when the service
+     * finished executing.
+     */
+    virtual int64_t service_stop_time();
+
+    virtual bool finished();
+
+    virtual void reset();
+
+private:
+
+    std::unique_ptr<leveldb::DB> metadata_db_;
+
+    int64_t service_start_time_;
```

Comment:
at some point we need to create this as a type we ship (along wiht other things like SN, etc). Is this something part of the reusability ideas you have?

Reply:
What type are your referring to?

Reply:
I was thinking to potentially create a struct that contains all these members, if these are going to be appearing in the implementation anyway. Something like ReplayTimeProperty or something along those lines